src/user_account/signals.py

```diff
@@ -0,0 +1,17 @@
+from django.db.models.signals import post_save, pre_save
+from django.dispatch import receiver
+
+from testset.models import TestResult
+from user_account.models import User
+
+
+@receiver(pre_save, sender=User)
+def save_profile(sender, instance, **kwargs):
+    count_passed_test = instance.count_passed_tests()
+    test_results = instance.test_results.filter(is_completed=True)
+
+    if count_passed_test != 0:
+        percent_success_passed = round((sum([test_result.correct_answers_count() for test_result in test_results]) / \
+                                        sum([test_result.test_question_count() for test_result in test_results]))
+                                       * 100, 2)
+        instance.percent_success = percent_success_passed
```

Comment:
Here is my implementation: https://github.com/dbradul/tmb/blob/master/src/account/signals.py

What is in my version:
1. Updated only when test result is finished, not when user is saved => it is guaranteed that recently finished test will be included into user's score.
2. I also optimized score update, so that it is calculated in one query (`User.update_score` method).
3. Some values are cached as models attributes (correct answers and total questions). I decided to display them in template to make it similar to test result and saving them as attributes allows to avoid queries when leaderboard is constructed. This is always a tradeoff between performance and data redundancy. In our case we have very small DB, so i only did it for demonstration. But potentially with huge number of users and test results retrieving this data might take some time.

Reply:
Thanks a lot!